Bug in cv::FileStorage when the XML contains integer values larger INT_MAX

### System Information

OpenCV version: 4.11.0
Operating System / Platform: Windows
Compiler: msvc 19.42.34436.0 (Visual Studio 2022)


### Detailed description

There seems to be a problem in the cv::FileStorage when it contains integers larger than INT_MAX.

When reading the Name node from an XML file like this one below, we get an error in modules\core\src\persistence.cpp:2233.
The XML reads fine for OpenCV up to version 4.10.0, or with integer values smaller than INT_MAX.

```xml
<?xml version="1.0"?>
    <opencv_storage>
      <A>
        <Value>2147483648</Value>
        <Name>"image"</Name>
      </A>
    </opencv_storage>
```

Output built with OpenCV 4.11.0
```
a is ok
b is ok
OpenCV(4.11.0) modules\core\src\persistence.cpp:2233: error: (-215:Assertion failed) key2 < fs->str_hash_data.size() in function 'cv::FileNode::operator []'
```

Output build with OpenCV 4.10.0
```
a is ok
b is ok
c is ok
```


### Steps to reproduce

```cpp
#include <iostream>
#include <opencv2/core/persistence.hpp>

int main() {
    auto example_a = R"(<?xml version="1.0"?>
    <opencv_storage>
      <A>
        <Value>2147483648</Value>
      </A>
    </opencv_storage>)";

    auto example_b = R"(<?xml version="1.0"?>
    <opencv_storage>
      <A>
        <Value>2147483647</Value>
        <Name>"image"</Name>
      </A>
    </opencv_storage>)";

    auto example_c = R"(<?xml version="1.0"?>
    <opencv_storage>
      <A>
        <Value>2147483648</Value>
        <Name>"image"</Name>
      </A>
    </opencv_storage>)";

    std::string name;
    cv::FileStorage fs;

    try {
        fs = cv::FileStorage(example_a, cv::FileStorage::READ | cv::FileStorage::MEMORY | cv::FileStorage::FORMAT_XML);
        fs["A"]["Name"] >> name;
        std::cout << "a is ok" << std::endl;
    } catch (std::exception &e) {
        std::cout << e.what() << std::endl;
    }

    try {
        fs = cv::FileStorage(example_b, cv::FileStorage::READ | cv::FileStorage::MEMORY | cv::FileStorage::FORMAT_XML);
        fs["A"]["Name"] >> name;
        std::cout << "b is ok" << std::endl;
    } catch (std::exception &e) {
        std::cout << e.what() << std::endl;
    }

    try {
        fs = cv::FileStorage(example_c, cv::FileStorage::READ | cv::FileStorage::MEMORY | cv::FileStorage::FORMAT_XML);
        fs["A"]["Name"] >> name;
        std::cout << "c is ok" << std::endl;
    } catch (std::exception &e) {
        std::cout << e.what() << std::endl;
    }

    return 0;
}

```

### Issue submission checklist

- [x] I report the issue, it's not a question
- [x] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [x] I updated to the latest OpenCV version and the issue is still there
- [x] There is reproducer code and related data files (videos, images, onnx, etc)

--- a/modules/core/src/persistence.cpp
+++ b/modules/core/src/persistence.cpp
@@ -2441,7 +2441,7 @@ size_t FileNode::rawSize() const
         p += 4;
     size_t sz0 = (size_t)(p - p0);
     if( tp == INT )
-        return sz0 + 4;
+        return sz0 + 8;
     if( tp == REAL )
         return sz0 + 8;
     if( tp == NONE )
@@ -2475,13 +2475,7 @@ void FileNode::setValue( int type, const void* value, int len )
         sz += 4;
 
     if( type == INT )
-    {
-        int64_t ival = *(const int64_t*)value;
-        if (ival > INT_MAX || ival < INT_MIN)
-            sz += 8;
-        else
-            sz += 4;
-    }
+        sz += 8;
     else if( type == REAL )
         sz += 8;
     else if( type == STRING )
@@ -2502,10 +2496,7 @@ void FileNode::setValue( int type, const void* value, int len )
     if( type == INT )
     {
         int64_t ival = *(const int64_t*)value;
-        if (sz > 8)
-            writeInt(p, ival);
-        else
-            writeInt(p, static_cast<int>(ival));
+        writeInt(p, ival);
     }
     else if( type == REAL )
     {

--- a/modules/core/test/test_io.cpp
+++ b/modules/core/test/test_io.cpp
@@ -2006,64 +2006,58 @@ TEST(Core_InputOutput, FileStorage_invalid_attribute_value_regression_25946)
 // see https://github.com/opencv/opencv/issues/26829
 TEST(Core_InputOutput, FileStorage_int64_26829)
 {
-    const std::string filename = cv::tempfile("FileStorage_int64_26829_test.xml");
+    String content =
+        "%YAML:1.0\n"
+        "String1: string1\n"
+        "IntMin: -2147483648\n"
+        "String2: string2\n"
+        "Int64Min: -9223372036854775808\n"
+        "String3: string3\n"
+        "IntMax: 2147483647\n"
+        "String4: string4\n"
+        "Int64Max: 9223372036854775807\n"
+        "String5: string5\n";
+
+    FileStorage fs(content, FileStorage::READ | FileStorage::MEMORY);
 
     {
-        FileStorage fs(filename, FileStorage::WRITE);
+        std::string str;
 
-        fs << "String1" << "string1";
-        fs << "IntMin" << INT_MIN;
-        fs << "String2" << "string2";
-        fs << "Int64Min" << INT64_MIN;
-        fs << "String3" << "string3";
-        fs << "IntMax" << INT_MAX;
-        fs << "String4" << "string4";
-        fs << "Int64Max" << INT64_MAX;
-        fs << "String5" << "string5";
+        fs["String1"] >> str;
+        EXPECT_EQ(str, "string1");
+
+        fs["String2"] >> str;
+        EXPECT_EQ(str, "string2");
+
+        fs["String3"] >> str;
+        EXPECT_EQ(str, "string3");
+
+        fs["String4"] >> str;
+        EXPECT_EQ(str, "string4");
+
+        fs["String5"] >> str;
+        EXPECT_EQ(str, "string5");
     }
 
     {
-        FileStorage fs(filename, FileStorage::READ);
+        int value;
 
-        {
-            std::string str;
+        fs["IntMin"] >> value;
+        EXPECT_EQ(value, INT_MIN);
 
-            fs["String1"] >> str;
-            EXPECT_EQ(str, "string1");
-
-            fs["String2"] >> str;
-            EXPECT_EQ(str, "string2");
-
-            fs["String3"] >> str;
-            EXPECT_EQ(str, "string3");
-
-            fs["String4"] >> str;
-            EXPECT_EQ(str, "string4");
-
-            fs["String5"] >> str;
-            EXPECT_EQ(str, "string5");
-        }
-
-        {
-            int value;
-
-            fs["IntMin"] >> value;
-            EXPECT_EQ(value, INT_MIN);
-
-            fs["IntMax"] >> value;
-            EXPECT_EQ(value, INT_MAX);
-        }
+        fs["IntMax"] >> value;
+        EXPECT_EQ(value, INT_MAX);
+    }
 
 
-        {
-            int64_t value;
+    {
+        int64_t value;
 
-            fs["Int64Min"] >> value;
-            EXPECT_EQ(value, INT64_MIN);
+        fs["Int64Min"] >> value;
+        EXPECT_EQ(value, INT64_MIN);
 
-            fs["Int64Max"] >> value;
-            EXPECT_EQ(value, INT64_MAX);
-        }
+        fs["Int64Max"] >> value;
+        EXPECT_EQ(value, INT64_MAX);
     }
 }
 

--- a/modules/core/test/test_io.cpp
+++ b/modules/core/test/test_io.cpp
@@ -2003,6 +2003,70 @@ TEST(Core_InputOutput, FileStorage_invalid_attribute_value_regression_25946)
     ASSERT_EQ(0, std::remove(fileName.c_str()));
 }
 
+// see https://github.com/opencv/opencv/issues/26829
+TEST(Core_InputOutput, FileStorage_int64_26829)
+{
+    const std::string filename = cv::tempfile("FileStorage_int64_26829_test.xml");
+
+    {
+        FileStorage fs(filename, FileStorage::WRITE);
+
+        fs << "String1" << "string1";
+        fs << "IntMin" << INT_MIN;
+        fs << "String2" << "string2";
+        fs << "Int64Min" << INT64_MIN;
+        fs << "String3" << "string3";
+        fs << "IntMax" << INT_MAX;
+        fs << "String4" << "string4";
+        fs << "Int64Max" << INT64_MAX;
+        fs << "String5" << "string5";
+    }
+
+    {
+        FileStorage fs(filename, FileStorage::READ);
+
+        {
+            std::string str;
+
+            fs["String1"] >> str;
+            EXPECT_EQ(str, "string1");
+
+            fs["String2"] >> str;
+            EXPECT_EQ(str, "string2");
+
+            fs["String3"] >> str;
+            EXPECT_EQ(str, "string3");
+
+            fs["String4"] >> str;
+            EXPECT_EQ(str, "string4");
+
+            fs["String5"] >> str;
+            EXPECT_EQ(str, "string5");
+        }
+
+        {
+            int value;
+
+            fs["IntMin"] >> value;
+            EXPECT_EQ(value, INT_MIN);
+
+            fs["IntMax"] >> value;
+            EXPECT_EQ(value, INT_MAX);
+        }
+
+
+        {
+            int64_t value;
+
+            fs["Int64Min"] >> value;
+            EXPECT_EQ(value, INT64_MIN);
+
+            fs["Int64Max"] >> value;
+            EXPECT_EQ(value, INT64_MAX);
+        }
+    }
+}
+
 template <typename T>
 T fsWriteRead(const T& expectedValue, const char* ext)
 {

